### PR TITLE
resource_manager: supports burstable mode

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve?

Issue Number: ref #4399 ref https://github.com/pingcap/tidb/issues/40380

### What is changed and how does it work?

```commit-message
resource_manager: supports burstable mode
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/mcs/resource_manager/client/client.go [hunk 1]
```diff
@@ -351,6 +352,9 @@ type groupCostController struct {
 		consumption *rmpb.Consumption
 	}
 
+	// fast path to make once token limit with un-limit burst.
+	burstable *atomic.Bool
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/client/client.go:356]: It seems that the standard library has an `atomic.Bool` type, can we use that one?

https://pkg.go.dev/sync/atomic#Bool

Comment (nolouch) [reply pkg/mcs/resource_manager/client/client.go:356]: addressed

File: pkg/mcs/resource_manager/client/client.go [hunk 2]
```diff
@@ -764,7 +784,9 @@ func (gc *groupCostController) onResponse(req RequestInfo, resp ResponseInfo) {
 	for _, calc := range gc.calculators {
 		calc.AfterKVRequest(delta, req, resp)
 	}
-
+	if gc.burstable.Load() {
+		return
```
Comment (CabinfeverB) [inline pkg/mcs/resource_manager/client/client.go:788]: we can goto `ret` like handle in `onRequestWait`

**Review by CabinfeverB — COMMENTED** [review]
rest LGTM

File: pkg/mcs/resource_manager/client/client.go [hunk 3]
```diff
@@ -516,21 +521,31 @@ func (gc *groupCostController) handleTokenBucketTrickEvent(ctx context.Context)
...
+		}
 		log.Debug("[resource group controller] update avg raw resource per sec", zap.String("name", gc.Name), zap.String("type", rmpb.RawResourceType_name[int32(typ)]), zap.Float64("avgRUPerSec", counter.avgRUPerSec))
 	}
+	gc.burstable.Store(isBurstable)
 }
 
 func (gc *groupCostController) updateAvgRUPerSec() {
+	isBurstable := true
 	for typ, counter := range gc.run.requestUnitTokens {
+		if counter.limiter.GetBurst() >= 0 {
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/client/client.go:540]: Should we place this check before or after the below check? I noticed that its place is different with `updateAvgRaWResourcePerSec`.

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

File: pkg/mcs/resource_manager/client/client.go (hunk 3 above)
Comment (nolouch) [reply pkg/mcs/resource_manager/client/client.go:540]: Good Catch, it's need before the below check.

**Review by CabinfeverB — APPROVED** [review]

**Comment by nolouch** [comment]
ptal @JmPotato

**Review by JmPotato — APPROVED** [review]

**Pull request merged**